### PR TITLE
[RUST] Write List/LargeList/FixedSizeList/FixedSizeBinary

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/src/io/reader.rs [hunk 1]
```diff
@@ -234,7 +235,23 @@ impl<'a> FileReader<'a> {
         let offset_arr = subtract_scalar(positions, start_position)?;
         let value_arrs = self.read_array(&field.children[0], batch_id).await?;
 
-        Ok(Arc::new(ListArray::new(value_arrs, &offset_arr)?))
+        Ok(Arc::new(ListArray::try_new(value_arrs, &offset_arr)?))
+    }
+
+    // TODO: merge with [read_list_array]?
```
Comment (changhiskhan) [inline rust/src/io/reader.rs:241]: could potentially create a read_generic_list_array<T> that both methods call with Int32 or Int64?

File: rust/src/io/writer.rs [hunk 2]
```diff
@@ -221,6 +228,26 @@ impl<'a> FileWriter<'a> {
...
+        assert!(!offsets.is_empty());
+        let offsets = Arc::new(subtract_scalar(&offsets, offsets.value(0))?) as ArrayRef;
+        self.write_fixed_stride_array(field, &offsets).await?;
+        self.write_array(&field.children[0], &list_arr.values())
+            .await
+    }
+
+    async fn write_large_list_array(&mut self, field: &Field, array: &ArrayRef) -> Result<()> {
+        let list_arr = as_large_list_array(array);
+        let offsets: Int64Array = list_arr.value_offsets().iter().map(|o| *o).collect();
```
Comment (changhiskhan) [inline rust/src/io/writer.rs:243]: does it not work if you just do `let offsets: &Int64Array = list_arr.value_offsets()` or something?

**Review by changhiskhan — APPROVED** [review]

File: rust/src/io/reader.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/io/reader.rs:241]: Yes, trying to use int32/int64 as template parameter to see how to get it working.

File: rust/src/io/writer.rs (hunk 2 above)
Comment (eddyxu) [reply rust/src/io/writer.rs:243]: `list_arr.value_offsets()` is a `&[i32]` array, not an Arrow Array.

Maybe i can just write the `&[i32]` array to disk directly?

Comment (eddyxu) [reply rust/src/io/writer.rs]: Oh, so offsets need to be computed to be relative offsets within the buffer again, which occurs a copy later.

**Pull request merged**